ethapi/api_test.go

```
@@ -596,6 +598,325 @@ func TestTransactionJSONSerialization(t *testing.T) {
+				Return(mockState, &header, nil).AnyTimes()
+			backend.EXPECT().GetEVM(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+				DoAndReturn(makeTestEVM(test.features)).AnyTimes()
+			backend.EXPECT().CurrentBlock().AnyTimes().Return(&evmcore.EvmBlock{EvmHeader: header})
+			backend.EXPECT().ChainConfig().AnyTimes().Return(makeChainConfig(test.features))
+			backend.EXPECT().SuggestGasTipCap(gomock.Any(), gomock.Any()).AnyTimes().Return(big.NewInt(1))
+			backend.EXPECT().MinGasPrice().AnyTimes().Return(big.NewInt(1))
+			backend.EXPECT().RPCGasCap().AnyTimes().Return(uint64(10000000))
+			backend.EXPECT().MaxGasLimit().AnyTimes().Return(uint64(10000000))
+			backend.EXPECT().StateAndHeaderByNumberOrHash(gomock.Any(), gomock.Any()).
+				Return(mockState, &header, nil).AnyTimes()
+			if test.extraSetupBackend != nil {
+				test.extraSetupBackend(backend)
+			}
+
+			// every test uses the same transaction
+			nonce := hexutil.Uint64(0)
+			gas := hexutil.Uint64(150_000)
+			gasFeeCap := hexutil.Big(*big.NewInt(10000000))
+			txArgs := TransactionArgs{
+				From:         &sender,
+				To:           &recipient,
+				Nonce:        &nonce,
+				Gas:          &gas,
+				MaxFeePerGas: &gasFeeCap,
+			}
+
+			test.call(t, backend, txArgs, blockOrHash)
+		})
+	}
+}
+
+// makeChainConfig allows to create a chain config with a given set of features
+func makeChainConfig(features opera.FeatureSet) *params.ChainConfig {
+	chainConfig := opera.MainNetRules().EvmChainConfig(
+		[]opera.UpgradeHeight{})
+
+	if features == opera.SonicFeatures {
+		chainConfig = opera.MainNetRules().EvmChainConfig(
+			[]opera.UpgradeHeight{
+				{Upgrades: opera.SonicFeatures.ToUpgrades(), Height: 0},
+			})
+	} else if features == opera.AllegroFeatures {
+		chainConfig = opera.MainNetRules().EvmChainConfig(
+			[]opera.UpgradeHeight{
+				{Upgrades: opera.SonicFeatures.ToUpgrades(), Height: 0},
+				{Upgrades: opera.AllegroFeatures.ToUpgrades(), Height: 1},
+			})
+	}
```

nit: silently returning config with zero value of `Upgrade` seems like not ideal for testing of future upgrades. What about `else{ fmt.Printf("unknown feature set %v", feature) }` ? or passing `t` as a parameter and using `t.Logf`?

---

https://github.com/0xsoniclabs/sonic/pull/146/commits/6c952c26bc125322cbfa1679dca2bc962a9a7d0a